torch/optim/lars.py

```diff
@@ -0,0 +1,168 @@
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+        for group in self.param_groups:
+            group.setdefault("nesterov", False)
+            group.setdefault("maximize", False)
+            group.setdefault("differentiable", False)
+
+    @_use_grad_for_differentiable
+    def step(self, closure=None):
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            params_with_grad = []
+            grads = []
+            momentum_buffer_list = []
+
+            for p in group["params"]:
+                if p.grad is not None:
+                    params_with_grad.append(p)
+                    grads.append(p.grad)
+
+                    state = self.state[p]
+
+                    momentum_buffer_list.append(state.get("momentum_buffer"))
+
+            lars(
+                params_with_grad,
+                grads,
+                momentum_buffer_list,
+                lr=group["lr"],
+                weight_decay=group["weight_decay"],
+                momentum=group["momentum"],
+                dampening=group["dampening"],
+                nesterov=group["nesterov"],
+                eps=self.eps,
+                trust_coefficient=self.trust_coefficient,
+                maximize=group["maximize"],
+            )
+
+            for p, momentum_buffer in zip(params_with_grad, momentum_buffer_list):
+                state = self.state[p]
+                state["momentum_buffer"] = momentum_buffer
+
+        return loss
+
+
+def lars(
+    params: List[Tensor],
+    grads: List[Tensor],
+    momentum_buffer_list: List[Optional[Tensor]],
+    *,
+    lr: float,
+    weight_decay: float,
+    momentum: float,
+    dampening: float,
+    nesterov: bool,
+    eps: float,
+    trust_coefficient: float,
+    maximize: bool,
+):
+    if not torch.jit.is_scripting():
+        func = _single_tensor_lars
+
+    func(
+        params,
+        grads,
+        momentum_buffer_list,
+        lr=lr,
+        weight_decay=weight_decay,
+        momentum=momentum,
+        dampening=dampening,
+        nesterov=nesterov,
+        eps=eps,
+        trust_coefficient=trust_coefficient,
+        maximize=maximize,
+    )
+
+
+def _single_tensor_lars(
+    params: List[Tensor],
+    grads: List[Tensor],
+    momentum_buffer_list: List[Optional[Tensor]],
+    *,
+    lr: float,
+    weight_decay: float,
+    momentum: float,
+    dampening: float,
+    nesterov: bool,
+    eps: float,
+    trust_coefficient: float,
+    maximize: bool,
+):
+    for i, param in enumerate(params):
+        d_p = grads[i] if not maximize else -grads[i]
+
+        p_norm = torch.norm(param.data)
+        g_norm = torch.norm(d_p.data)
+
+        if weight_decay != 0:
+            # LARS scaling:
+            if p_norm * g_norm > 0:
+                lars_lr = trust_coefficient * p_norm / (g_norm + p_norm * weight_decay + eps)
```

Comment:
Not sure the eps is required: if p_norm is zero, then the whole thing is zero, if g_norm is zero, then the basic update term is zero (same effect as using local LR = 1 if any of those is zero, but we avoid the eps imprecision :))

Reply:
Yes, but you forgeot the case when both are zero!

Reply:
p_norm and g_norm are banned from being 0 with the conditional, no?

That said, I am not sure why lightning included an eps.

Reply:
```suggestion
                lars_lr = trust_coefficient * p_norm / max(g_norm + p_norm * weight_decay, eps)
```
This would avoid unnecessarily biasing the denominator.

Reply:
I checked the code again, and I think you're right.